### DISCUSSION TITLE
SFTP Reports Scheduling Feature

### DISCUSSION THREAD

**Discussion by rohitpawar2811** [apache/superset#24296] [Ideas]
**Proposing the SFTP Scheduling Feature in Superset**

I am about to contribute SFTP-Scheduling support feature currently working on it.

I have thoughts to make this feature like :

1. This one is just like Email, one SFTP is there which is configured through config-properties, and in UI you can simply select SFTP Method, and in the recipient section, we can give the target location of sftp to place the file. : Tested this idea and it is working.

![Screenshot from 2023-06-07 13-12-38](https://github.com/apache/superset/assets/72196393/2fc5f1eb-73cd-49a4-ad4f-652a5a1bb27a)

2. It's Totally different scenario In which, There are Multiple SFTPs that can be configured like databases, and on alert/reports UI on the selection of SFTP notification method we would show a selection box for available SFTP connections and the recipient things same as above **I need your suggestion on this path**

- In a second way, we would get the flexibility to use different-different SFTPs not stick to just one, and It will enable us to work in
multi-tenant scenarios.

- We can not deal with one SFTP with 100 or more users because SFTP has a limit on connection, R/W users, and concurrent access by 100 users which will lead us to a connection failure problem.

**I want your thoughts or suggestion on which way we have to go?, Or anyone has any Idea regards to the 2nd scenario?**

@eschutho
@rusackas
@betodealmeida

**Comment by rohitpawar2811** [comment]
I have thoughts to make this feature like :

1. Just like Email, one SFTP is there and in the recipient section, we can give target-location of sftp to place the file. : Raw tested this idea and it is working.

2. It's Totally different scenario In which, There are Multiple SFTPs that can be configured like databases, and on alert/reports UI on the selection of SFTP notification method we would show a selection box for available SFTP connections and the recipient things same as above.  :  **I need your suggestion on this path**

- In a 2nd way, we would get the flexibility to use different -different SFTPs not stick to just one, and It will enable us to work in 
**multi-tenant** scenarios 
- We can not deal with one SFTP with 100 or more users because SFTP has a limit on connection, R/W users, and concurrent access by 1k users that will lead us to a connection failure problem.

**I want your thoughts or suggestion on which way we have to go?, Or anyone has any Idea regards to the 2nd scenario?** 

@eschutho 
@rusackas 
@betodealmeida

**Comment by rusackas** [comment]
If #1 works, and doesn't present any major issues, I would just go that route. 

#2 sounds like it would be a lot more intricate, as this would open the UI to accessing buckets with write access. This opens a whole world of complexity around RBAC for the UI and for adding/editing/deleting the connections/sources. It'll also open edge cases if reports point to data sources that get changed/deleted, etc. 

I'm also curious a couple key things:
1) What are the use cases, i.e. what file structure(s) will you be storing on the SFTP server, and what will be consuming them?
2) What alternatives were considered, e.g. other file storage systems like S3, or APIs for things like Dropbox/Box/GoogleDrive/etc? Do these help solve the concurrency limit?

**Comment by rohitpawar2811** [comment]
1. Our use case is that we need the data feed for our integration systems, so we are using superset to generate data feeds and place it on any file-storage system (currently using SFTP as it is widely used ) and our 3rd party integration system picks data from there

2. Yes we can go for S3, can we configure S3 in superset to upload reports to s3 buckets? How?
>> Dropbox/Box/GoogleDrive/

and Can we use these options to upload data on a regular basis?

**Reply by rusackas** [reply]
Is it that you're trying to upload chart images, or are you trying to upload data (CSV, etc.)? I don't know what the integration system does, so I'm still not sure of the use case, or the results/formats needed to feed it.

Superset does not have S3 or other SaaS tool API services. I'm just wondering if any of them make as much (or more) sense than SFTP, given the limitations you mentioned.

**Reply by rohitpawar2811** [reply]
I am looking for uploading chart data in CSV Format yet.

**Reply by rohitpawar2811** [reply]
>> Superset does not have S3 or other SaaS tool API services. I'm just wondering if any of them make as much (or more) sense than SFTP, given the limitations you mentioned

In S3 there is no limit on connection but still we want to manage our customer reports in separate storage,for management and better permission access. and other reason is, it might happen that per customer we have to handle large amount of files

**Comment by billbelanger** [comment]
Just adding a +1 for the cloud delivery options. Especially if we can have multiple delivery options. In our case we'd like to be able to deliver customer reports directly to their S3 boxes (or ones that we've set up for them).

**Comment by archleadergt** [comment]
Just adding a +1 for the on premise sftp and cloud delivery options. Some companies would need send on premise files and others cloud delivery options
